crates/matrix-sdk/src/widget_api/capabilities.rs

```diff
@@ -1,70 +1,50 @@
 use async_trait::async_trait;
+use tokio::sync::mpsc::Receiver;
 
 use super::{
     messages::{
-        capabilities::{EventFilter, Options},
-        MatrixEvent, from_widget::{SendEventResponse, SendToDeviceRequest},
+        capabilities::Options,
+        from_widget::{SendEventResponse, SendToDeviceRequest, ReadEventRequest, SendEventRequest, ReadEventResponse},
+        MatrixEvent,
     },
     Result,
 };
 
 /// A wrapper for the matrix client that only exposes what is available through the capabilities.
 #[allow(missing_debug_implementations)]
 pub struct Capabilities {
-    pub event_reader: Option<Box<dyn EventReader>>,
-    pub event_writer: Option<Box<dyn EventWriter>>,
+    // The options are a private member,
+    // they contain the type filters for the listeners.
+    options: Options,
+
+    pub room_event_listener: Option<Box<Receiver<MatrixEvent>>>,
+    pub state_event_listener: Option<Box<Receiver<MatrixEvent>>>,
```

Comment:
A couple of notes:
1. We don't need to box the receiver, because it's sendable as long as the event is sendable.
2. Since we use 2 different receivers, we need 2 different types, i.e. a `StateEvent` and a `TimelineEvent`.

Reply:
Of course this was copied from the EventSender/EventReder without properly thinking about it.

Reply:
I think the matrix event in this case is generic enough (with the optional state key) to allow both types. So whatever I send you here should be possible to go into the content of the accoding ToWidget::read_event message.

Reply:
```
pub struct MatrixEvent {
    #[serde(rename = "type")]
    event_type: String,
    sender: String,
    event_id: String,
    room_id: String,
    state_key: Option<String>,
    origin_server_ts: u32,
    content: serde_json::Value,
    unsigned: Unsigned,
}
```
This should cover all possible events sent over the widget api

Reply:
Yeah, but the question is if we should allow sending state events to the function that does not mean accepting state events. Otherwise the receiver will have to check for the `state_key`.

Reply:
I am convinced! Will change that

Reply:
Following the code paths, it looks like differenciating between state and room events is not our business.
Especially with the matrosca mode we pretend that the widget driver can be treated like a matrix server. So we also want to get the same error responses if we send a malformatted event. Sending a room event with a state key should not be caught by the widget driver but instead should go to the homeserver and the homeserver should provide the response or treat the event the same it would if the widget would not be used in matroska mode.

Reply:
basically we really dont care if it is a state or room event and never look inside the event. Maybe even parsing the event is already too much, but we need it to evaluate the permissions/capabilities. Everything else is not our business since we are really just a gate for the widget to communicate with the homeserver. (and beeing a gate should not add additional checks except the ones necassary for checking permissions)

Reply:
But there was another bug i found. the state key needs to be optional!
Otherwise the widget dirver would always only send state events with empty keys... That would be a bummer

Reply:
Regarding the widget types. The main reason I suggested we have different types is to ensure that you can't send a state event to the timeline event sender or a timeline event to a state event sender. The only way to ensure it on compile time is to have different types. An alternative is to force the receiver of the message to check for the `state` field, but this would eliminate the rationale of having separate channels for state/timeline events in the first place 🙂

Reply:
Currently the widget does not provide a dedicated action for state events. So we need to evaluate if the message contains a state_key or not anyways to decide on which sender we use to send it.
But you are right it somewhat eleminates the channels but for the capability (permissions) we again have seperate filters.
So we have two capability filter arrays: one for state and one for room events. We get a send event from the widget and at some point need to decide what it is and than apply the correct filters.
I had in mind this decision is done in the handler and than the correct sender is used so that the right filter array is applied. wdyt?
From a bug likelyhood perspective the best solution would be to drift away from the widget api and have two different incoming messages based on the availability of the state_key field (that can be done with serde very nicely with `#[serde(untagged)]`) than we have different types from the getgo and always tread them seperatly without any errors possible if we always make room and state event types incompatible.

Reply:
> Currently the widget does not provide a dedicated action for state events. So we need to evaluate if the message contains a state_key or not anyways to decide on which sender we use to send it.
> But you are right it somewhat eleminates the channels but for the capability (permissions) we again have seperate filters.

Right, then it's a question of: do we really need to use 2 channels then? The thing is that if the event that we send has an optional field and this field must be checked by the receiver for correctness anyway, then there is no need for separate channels. I.e.: we either need 2 channels with 2 distinct types, this way we can't send a wrong thing from the handler. Or we need 1 channel with 1 "super-type" (used for both state/timeline events) and do the checking on the receiver's side. Since I'm not very familiar with how often/useful these are for the Widget API, I would be fine with both options I think 🙂 However the current option is using 2 channels with 1 data type which does not correspond to any of those.

Reply:
one sender/listener seems to be what we want it to be.